core/translate/optimizer.rs

```diff
@@ -9,35 +9,57 @@ use super::plan::{
     Direction, IterationDirection, Plan, Search, SourceOperator,
 };
 
+pub fn optimize_select_plan(plan: Plan) -> Result<Plan> {
+    optimize_plan(plan, true, true, true)
+}
+
+pub fn optimize_delete_plan(plan: Plan) -> Result<Plan> {
+    optimize_plan(plan, false, true, false)
+}
+
 /**
  * Make a few passes over the plan to optimize it.
  * TODO: these could probably be done in less passes,
  * but having them separate makes them easier to understand
  */
-pub fn optimize_plan(mut select_plan: Plan) -> Result<Plan> {
+fn optimize_plan(
+    mut select_plan: Plan,
```

Comment:
this shouldn't be called `select_plan` anymore if we are sharing the code.

however i think it would be better to just duplicate code in `optimize_select_plan` and `optimize_delete_plan` and call the functions that each of those want to call, instead of passing a bunch of bools around.

Reply:
duplicating seems better choice for now. 👍